### PR TITLE
Add programmatic API

### PR DESCRIPTION
**Pull request by vdemedes**
This PR adds a long-awaited programmatic API (#83).

To-do:

- [x] resolve errors on node v0.10
- [ ] add tests
- [ ] use API in tests instead of `child_process.exec`
- [x] implement #282

It also opens a road for #27 (TAP support).

### CODE REVIEW

**Comment by jamestalmage** [comment]
I think this should go in the base directory:

```js
require('ava/api');
// is better than
require('ava/lib/api');
```

File: lib/api.js [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+	this.passCount = 0;
+	this.failCount = 0;
+	this.fileCount = 0;
+	this.testCount = 0;
+	this.errors = [];
+	this.stats = [];
+	this.tests = [];
+
+	Object.keys(Api.prototype).forEach(function (key) {
+		this[key] = this[key].bind(this);
```
Comment (jamestalmage) [inline lib/api.js]: Maybe add a typeof check?
```js
if (typeof this[key] === 'function') {
  this[key] = this[key].bind(this);
}
```

File: lib/api.js [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+	if (isError) {
+		this.errors.push(test);
+	} else {
+		test.error = null;
+	}
+
+	this.emit('test', test);
+};
+
+Api.prototype._prefixTitle = function (file) {
```
Comment (jamestalmage) [inline lib/api.js]: `_prefixTitle` introduces new behavior not related to exposing an API. I like where you are headed, but it should probably be a separate PR.

File: index.js [hunk 3]
```diff
@@ -66,8 +66,17 @@ function exit() {
...
 globals.setImmediate(function () {
+	send('stats', {
+		testCount: runner.select({type: 'test'}).length
+	});
+
 	runner.on('test', test);
-	runner.run().then(exit);
+
+	process.on('message', function (message) {
+		if (message.name === 'ava-run') {
```
Comment (jamestalmage) [inline index.js]: Just use:

```js
process.on('ava-run', ...);
```

File: lib/api.js [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+			var statsCount = 0;
+			var deferred = Promise.pending();
+
+			tests.forEach(function (test) {
+				test.on('stats', function () {
+					if (++statsCount === self.fileCount) {
+						self.emit('ready');
+
+						tests.forEach(function (test) {
+							test.run();
```
Comment (jamestalmage) [inline lib/api.js]: Looks like `test.run()` gets called a bunch of times.

Comment (jamestalmage) [inline lib/api.js]: Why are you waiting for every test file to load before kicking off the tests? Once an individual test is `ready`, why not let it start?

**Comment by jamestalmage** [comment]
@vdemedes - see my change. 

The issue you were having was `stats` never gets emitted if there is syntax error in the test file (that is why it started failing on the `babel-hook` fixture, which intentionally has ES2015 syntax in a file that will not be instrumented by babel).

**Comment by jamestalmage** [comment]
I propose we change this up just a bit.

Rather than `api` tracking counts for `unhandledRejection`, `uncaughtException`, etc.

Use the array: `api.errors`, and every time one is encountered, push to it:

```js
on('unhandledRejection', function(e) {
  self.errors.push({type: 'unhandledRejection', error: e});
});

on('uncaughtException', function(e) {
  self.errors.push({type: 'uncaughtException', error: e});
});

// ...
self.errors.push({
  type: 'nonZeroExit',
  error: new Error('process exited with code: ' + code)
});

// ...
self.errors.push({
  type: 'assertionError',
  // ...
});
```

Then extend API with methods like:
```js
api.hasErrors(); // true / false
api.getErrors(); // returns all {type: typeName, error: errorObj}
api.getErrors(typeName); // returns all errors of matching type. 
```

I think the promise returned by the API should almost always resolve, and almost never reject. Users will then use `hasErrors`, and decide what to do. This makes `unhandledRejection`, `uncaughtException`, and `nonZeroExit` consistent with what we do with a failed assertion (we resolve the promise with a result that indicates there were errors).

We should continue emitting events so people can implement their own realtime logging, but it should be possible to create basically the same log output after the fact using just the array returned from  `getErrors()`.

Thoughts?

**Comment by jamestalmage** [comment]
I've shared a number of thoughts, but overall, very much :+1:

**Comment by jamestalmage** [comment]
>  use API in tests instead of child_process.exec

I'm not sure of this one. I like that `test/cli.js` exercises the full stack. It currently acts as more of an integration test.

File: cli.js [hunk 5]
```diff
@@ -48,133 +43,60 @@ var cli = meow([
...
-
-	return prefix;
-}
+updateNotifier({pkg: cli.pkg}).notify();
 
-function stats(stats) {
-	testCount += stats.testCount;
+if (cli.flags.init) {
+	require('ava-init')();
+	process.exit();
```
Comment (sindresorhus) [inline cli.js]: You can't `process.exit()` here as `ava-init` won't have a time to finish. Just `return` and XO will pass when https://github.com/sindresorhus/ava/pull/277/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R132 lands.

File: cli.js [hunk 6]
```diff
@@ -48,133 +43,60 @@ var cli = meow([
...
-function test(data) {
-	var isError = data.error.message;
-
-	if (fileCount > 1) {
-		data.title = prefixTitle(data.file) + data.title;
-	}
-
-	if (isError) {
-		log.error(data.title, chalk.red(data.error.message));
+var api = new Api({
```
Comment (sindresorhus) [inline cli.js]: Maybe `new Api(files, opts)`?

File: lib/api.js (hunk 2 above)
Comment (sindresorhus) [inline lib/api.js]: @jamestalmage We already do this in https://github.com/sindresorhus/ava/blob/e2e45151dc5ac249d49a7cd94baca8c70f9ed2ce/cli.js#L57

File: lib/api.js [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+	var self = this;
+
+	return handlePaths(this.files)
+		.map(function (file) {
+			return path.resolve(file);
+		})
+		.then(function (files) {
+			if (files.length === 0) {
+				var err = new Error('Couldn\'t find any files to test');
+				return Promise.reject(err);
```
Comment (sindresorhus) [inline lib/api.js]: `return Promise.reject(new Error('Couldn\'t find any files to test'));`

File: lib/api.js [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+			var deferred = Promise.pending();
+
+			tests.forEach(function (test) {
+				var counted = false;
+
+				function count() {
+					if (counted) {
+						return;
+					}
+					if (++statsCount === self.fileCount) {
```
Comment (sindresorhus) [inline lib/api.js]: add an empty line above this

**Comment by sindresorhus** [comment]
> We might be able to move the majority of them to test/api.js, but some should definitely stay behind. 

@jamestalmage Yup, that's the plan. We still want to explicitly test the CLI, but many of the current CLI tests really just test "API" behavior.

**Comment by sindresorhus** [comment]
:+1: Overall looks really good!

File: lib/api.js (hunk 2 above)
Comment (jamestalmage) [inline lib/api.js]: so we do :+1:

File: lib/api.js (hunk 1 above)
Comment (vdemedes) [inline lib/api.js]: Copied from `lib/test.js`.

**Comment by vdemedes** [comment]
@jamestalmage I appreciate your suggestions, but I'd like to ask you not to modify WIP-kind of PRs. I already had a fix in development, but I had to cancel all my changes to merge yours in.

**Comment by jamestalmage** [comment]
Sure no - problem. Sorry for messing you up.

**Comment by vdemedes** [comment]
@jamestalmage Good suggestion about errors!

**Comment by vdemedes** [comment]
I'd really like to merge this now and add TAP support for the next release. I could add/convert tests in a separate PR. @sindresorhus @jamestalmage what do you think?

**Comment by jamestalmage** [comment]
I would wait for the release. We already have one major refactor in this release. I think that we let major changes sit on master for at least a few days before release.

**Comment by sindresorhus** [comment]
> I think that we let major changes sit on master for at least a few days before release.

I agree. Not really any user-facing changes here anyways.

**Comment by sindresorhus** [comment]
Landed! This is exactly what I had in mind when I opened the issue about it. Really nice work @vdemedes :)

**Comment by sindresorhus** [comment]
tumblr_mlf8zt33uq1qdlh1io1_400

**Comment by jamestalmage** [comment]
I need to step up my GIF game.

File: api.js [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+		.map(function (file) {
+			if (fs.statSync(file).isDirectory()) {
+				return handlePaths([path.join(file, '*.js')]);
+			}
+
+			return file;
+		})
+		.then(flatten)
+		.filter(function (file) {
+			return path.extname(file) === '.js' && path.basename(file)[0] !== '_';
```
Comment (azizhk) [inline api.js]: How do .jsx test files run?
@vdemedes 

Debugged to find this, but unfortunately/luckily this file has been refactored by @jamestalmage 
Hoping that its fixed in next release or minor release to patch this is released soon.

Comment (azizhk) [inline api.js]: Ok the code has been moved to https://github.com/avajs/ava/blob/e01ee00/lib/ava-files.js#L248
Will see if I can account extension from files or source.
